js/modules/gomodule.go

```diff
@@ -4,89 +4,99 @@ import (
+		}
+		for _, callback := range gm.exportedNamesCallbacks {
+			callback(gm.exportedNames)
+		}
 	}
-	return b.exportsO
+	return &goModuleInstance{rt: rt, mi: mi}, nil
 }
 
-// goModule is a go module which implements Module
-type goModule struct {
-	Module
-}
+func (gm *goModule) Evaluate(_ *sobek.Runtime) *sobek.Promise { panic("this shouldn't happen") }
 
-var _ module = &goModule{}
+func (gm *goModule) GetExportedNames(callback func([]string), _ ...sobek.ModuleRecord) bool {
+	if gm.exportedNames != nil {
+		callback(gm.exportedNames)
+		return true
+	}
+	gm.exportedNamesCallbacks = append(gm.exportedNamesCallbacks, callback)
+	return false
+}
 
-func (g *goModule) instantiate(vu VU) moduleInstance {
-	return &goModuleInstance{vu: vu, module: g}
+func (gm *goModule) ResolveExport(exportName string, _ ...sobek.ResolveSetElement) (*sobek.ResolvedBinding, bool) {
+	return &sobek.ResolvedBinding{
+		Module:      gm,
+		BindingName: exportName,
+	}, false
 }
 
 type goModuleInstance struct {
-	Instance
-	module   *goModule
-	vu       VU
-	exportsO *sobek.Object // this is so we only initialize the exports once per instance
+	mi            Instance
+	rt            *sobek.Runtime
+	defaultExport sobek.Value
 }
 
-var _ moduleInstance = &goModuleInstance{}
-
-func (gi *goModuleInstance) execute() error {
-	gi.Instance = gi.module.NewModuleInstance(gi.vu)
-	return nil
+func (gmi *goModuleInstance) ExecuteModule(_ *sobek.Runtime, _, _ func(any)) (sobek.CyclicModuleInstance, error) {
+	return gmi, nil
 }
+func (gmi *goModuleInstance) HasTLA() bool { return false }
 
-func (gi *goModuleInstance) exports() *sobek.Object {
-	if gi.exportsO == nil {
-		rt := gi.vu.Runtime()
-		gi.exportsO = rt.ToValue(toESModuleExports(gi.Instance.Exports())).ToObject(rt)
+func (gmi *goModuleInstance) GetBindingValue(name string) (v sobek.Value) {
+	if name == "default" {
+		return gmi.getDefaultExport()
 	}
-	return gi.exportsO
-}
 
-func toESModuleExports(exp Exports) interface{} {
-	if exp.Named == nil {
-		return exp.Default
+	exports := gmi.mi.Exports()
+	if exports.Named != nil {
+		return gmi.rt.ToValue(exports.Named[name])
 	}
-	if exp.Default == nil {
-		return exp.Named
+	return gmi.getDefaultExport().ToObject(gmi.rt).Get(name)
+}
+
+func (gmi *goModuleInstance) getDefaultExport() sobek.Value {
+	if gmi.defaultExport != nil {
+		return gmi.defaultExport
 	}
 
-	result := make(map[string]interface{}, len(exp.Named)+2)
+	exports := gmi.mi.Exports()
+	if exports.Default != nil {
+		gmi.defaultExport = gmi.rt.ToValue(exports.Default)
+		return gmi.defaultExport
+	}
 
-	for k, v := range exp.Named {
-		result[k] = v
+	// if there are only named exports we make a default object out of them
+	// this allows scripts to modify this acting similar to how it would act
+	// if the default export was an object to begin with.
+	o := gmi.rt.NewObject()
+	gmi.defaultExport = o
+	for name, value := range exports.Named {
+		// TODO:maybe do something slightly smarter
```

Comment:
I think I was hoping I can do something where we only import and set them if/when they are used. But that has pretty big problems in the general case where that might happen *after* init context, so I am just going to drop it.

Using field accessors doesn't seem like it will add any functionality here ... maybe for counting what is used, but that is a separate problem.